### ISSUE TITLE
[Proposal]: Improved interpolated strings

### ISSUE THREAD

**Issue by 333fred** [dotnet/csharplang#4487]
# Improved interpolated strings

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

Split out from https://github.com/dotnet/csharplang/issues/2302.

https://github.com/dotnet/csharplang/pull/4486

**Comment by yaakov-h** [comment]
I don't understand this part of the spec:

> By making the builder an `out` parameter we allow overloading of the `GetInterpolatedStringBuilder` method by builder type, which is useful for scenarios like the logger above, which could have `TraceLoggerParamsBuilder`/`DebugLoggerParamsBuilder`/`WarningLoggerParamsBuilder`/etc.

How would the overload occur in practice when the compiler translates the string interpolation expression using `out var` or equivalent?:

```csharp
TraceLoggerParamsBuilder.GetInterpolatedStringBuilder(baseLength: 47, formatHoleCount: 1, receiverTemp, out var builder);
```

How would the compiler know which overload to select if there were multiple? By my reading of the spec, `GetInterpolatedStringBuilder` would have to be implemented as a static method on 3 different types in order to have `TraceLoggerParamsBuilder`/`DebugLoggerParamsBuilder`/`WarningLoggerParamsBuilder` rather than being able to use an overload.

**Comment by 333fred** [comment]
@yaakov-h I used var for convenience there. It would not be var in practice.

**Comment by yaakov-h** [comment]
In that case how would the compiler ever resolve an overload to something other than the type itself?

> When an interpolated_string_expression is passed as an argument to a method, we look at **the type of the parameter**. If the parameter type has a static method GetInterpolatedStringBuilder ... and has an out parameter of **the type of original method's parameter**

It seems like those two types will always be the same, so where is there room for overloading?

Or would it be the case that a base class could provide overloads for a number of its descendant classes?

**Comment by 333fred** [comment]
> In that case how would the compiler ever resolve an overload to something other than the type itself?

I'm not sure what you mean. The idea is that a logging provider could provide `GetInterpolatedString(..., out TraceLoggerParamsBuilder/InfoLoggerParamsBuilder/etc builder)`, and then `LogTrace(TraceLoggerParamsBuilder builder)`, `LogInfo(InfoLoggerParamsBuilder builder)`, etc.

**Comment by yaakov-h** [comment]
I'm confused then.

My understanding based on my reading of the spec is that the _parameter type_ has to provide `GetInterpolatedString`, so you would need:

```csharp
ref struct TraceLoggerParamsBuilder
{
    public static void GetInterpolatedStringBuilder(..., out TraceLoggerParamsBuilder builder) { ... }
}

ref struct InfoLoggerParamsBuilder
{
    public static void GetInterpolatedStringBuilder(..., out InfoLoggerParamsBuilder builder) { ... }
}

ref struct WarnLoggerParamsBuilder
{
    public static void GetInterpolatedStringBuilder(..., out WarnLoggerParamsBuilder builder) { ... }
}

interface ILog
{
    void LogTrace(TraceLoggerParamsBuilder builder);
    void LogInfo(InfoLoggerParamsBuilder builder);
    void LogWarn(WarnLoggerParamsBuilder builder);
}
```

How could this be restructured such that _one class_ defines _all three_ overloads of `GetInterpolatedStringBuilder`?

**Comment by berkeleybross** [comment]
> How could this be restructured such that _one class_ defines _all three_ overloads of `GetInterpolatedStringBuilder`?

Could we you attributes instead?

```c#
ref struct LogParamsBuilder {
     public static InfoLoggerParamsBuilder GetTraceInterpolatedStringBuilder(...) { ... }
     public static InfoLoggerParamsBuilder GetInfoInterpolatedStringBuilder(...) { ... }
     public static InfoLoggerParamsBuilder GetWarnInterpolatedStringBuilder(...) { ... }
}


interface ILog
{
    void LogTrace([ParamBuilder("GetTraceInterpolatedStringBuilder")] LogParamsBuilder builder);
    void LogInfo([ParamBuilder("GetInfoInterpolatedStringBuilder")] LogParamsBuilder builder);
    void LogWarn([ParamBuilder("GetWarnInterpolatedStringBuilder")] LogParamsBuilder builder);
}
```

Where the attribute takes the method name of the type.
It could infer the type from the parameter type like the current proposal, or optionally take a typeof(LogParamsBuilder ) if for some reason the builder is different from the ParamsBuilder

Is there a performance gain by using an out parameter or was it just for overload resolution? It seems more intuitive to return a value then to have a void method with a single out parameter.

**Comment by berkeleybross** [comment]
Im not sure on how the compiler would behave here, but it seems easier for the compiler if is a parameter attribute. If its an arbitrary class as a parameter to a method, then presumably _every_ method call has to figure out whether the parameter is a possible param builder. Whereas the attribute will efficiently and obviously indicate this behaviour

**Comment by 333fred** [comment]
> then presumably every method call has to figure out whether the parameter is a possible param builder.

Yes, but only in cases where an interpolated string is actually passed as an argument, and only for that particular parameter type. This is not likely to be a very expensive check, I'm not worried about it :).

**Comment by ufcpp** [comment]
How can this treat item format?

```cs
Log($"{i:X4}");
```

**Comment by 333fred** [comment]
> How can this treat item format?

However the builder wants? It needs to have an overload of `TryFormat` that accepts the interpolation hole content, which is both `i` and the string `"X4"`. What happens after that is up to the builder type.

**Comment by ufcpp** [comment]
I want to avoid string allocation of `i.ToString("X4")`. As an alternative idea, I want `ISpanFormattable` to be public.

```cs
Span<char> buffer = stackalloc char[4];
((ISpanFormattablei)i).Format(buffer, out var written, "X4");
Log(buffer.AsSpan(0, written));
```

**Comment by yaakov-h** [comment]
Wouldn't the existing design avoid that kind of allocation if you chose to implement it kinda like this?:

```csharp
public bool TryFormat(int i, string format)
{
    Span<char> buffer = stackalloc char[4];
    bool formatted = i.Try.Format(buffer, out var written, format);
    Debug.Assert(formatted == true);

    Log(buffer.AsSpan(0, written));
}

**Comment by 333fred** [comment]
> Wouldn't the existing design avoid that kind of allocation if you chose to implement it kinda like this?:
> 
> ```csharp
> public bool TryFormat(int i, string format)
> {
>     Span<char> buffer = stackalloc char[4];
>     bool formatted = i.Try.Format(buffer, out var written, format);
>     Debug.Assert(formatted == true);
> 
>     Log(buffer.AsSpan(0, written));
> }

Yup. The builder implements the TryFormat method. We never call ToString on any components in this proposal.

**Comment by ufcpp** [comment]
I found spec https://github.com/dotnet/csharplang/pull/4486/files#diff-77eda16068bf940c41c3c540d5a0f74c1bfee4b56388b0bc4e59e037abaa9ec7R124-R125. This makes sense. Thank you.

**Comment by miyu** [comment]
Would implementing TryFormat automatically affect the default ToString's behavior?

```csharp
struct Point
{
    public int X, Y;
    public bool TryFormat(Span<char> destination, out in charsWritten, ReadOnlySpan<char> format, IFormatProvider? provider) =>
        destination.TryWrite($"X={X} Y={Y}", out charsWritten, provider);
    public override string ToString() => $"{this}"; // is this line necessary?
}
```

(Based on an example graciously provided by @stephentoub in https://github.com/dotnet/runtime/issues/50601)

**Comment by 333fred** [comment]
I don't particularly understand what you're asking @miyu. Presumably, the first parameter of `TryWrite` takes a builder type, and that would format the string however it wants.

**Comment by KalleOlaviNiemitalo** [comment]
<https://github.com/dotnet/csharplang/blob/d41fcd0377db1ecccfa8de9185b9aa07f894b342/proposals/improved-interpolated-strings.md#lowering> does not say how the compiler generates the `int baseLength, int formatHoleCount` arguments for the `Create` call. Is `baseLength` required to be exactly the total number of characters outside the interpolation holes, or is the compiler allowed to pad `baselength` to improve the chance that the initial buffer will be large enough? I imagine such padding could be beneficial if the compiler generates calls of the `public void TryFormatInterpolationHole<T>(T value, int alignment, string format)` method proposed in <https://github.com/dotnet/runtime/issues/50601> and notices an unusually large `alignment`.

**Comment by 333fred** [comment]
>  Is `baseLength` required to be exactly the total number of characters outside the interpolation holes, or is the compiler allowed to pad `baselength` to improve the chance that the initial buffer will be large enough?

The exact length. That's why both values are passed, so the underlying API can make its own padding decisions.

**Comment by stephentoub** [comment]
@KalleOlaviNiemitalo has a good point, though.  The compiler ends up potentially having more information it could pass to the Create method to enable the builder to do better presizing. The builder isn't currently told how big each hole is,  and so can only guess at some average, but the compiler may be given more info in advance,  e.g. an alignment that provides a minimum length on a particular hole's size. We should consider whether baseLength should instead be the minimum length the compiler guarantees will be required to format,  rather than it just being the number of characters in the base strings. (It must, however,  never overestimate,  or some builders will end up being functionally incorrect.)

I've also been wondering if the compiler might be able to help the builder track past state to inform future use.  Imagine if the builder was passed some kind of unique id per call site. It could then use that id if desired to remember the typical length of strings created for that call site in past executions in that process, in order to better presize in the Create method for future iterations.

**Comment by KalleOlaviNiemitalo** [comment]
I meant the spec doesn't have any wording to disallow a compiler from generating e.g. `InterpolatedStringBuilder.Create(-1, -1)` regardless of what is in the interpolated string expression. The argument values should be specified rather than only implied by parameter names.

**Comment by 333fred** [comment]
The lowering section still needs to be fleshed out, and that will be based on LDM feedback on other parts of the proposal.

**Comment by ufcpp** [comment]
It may already be too late, but can the parameter of `AppendLiteral` be `char` in addition to `string`?

e.g. now,

```cs
string s = $"{a}.{b}.{c}.{d}"
```

is lowered to

```cs
handler.AppendFormatted(a);
handler.AppendLiteral(".");
handler.AppendFormatted(b);
handler.AppendLiteral(".");
handler.AppendFormatted(c);
handler.AppendLiteral(".");
handler.AppendFormatted(d);
```

however can `handler.AppendLiteral(".")` be changed to `handler.AppendLiteral('.')` if the literal is a single character and the handler has a `char` overload?

**Comment by stephentoub** [comment]
> can the parameter of AppendLiteral be char in addition to string?

Is your concern performance?  It won't matter with https://github.com/dotnet/runtime/pull/57217.

**Comment by ufcpp** [comment]
> Is your concern performance?

Yes.

Is https://github.com/dotnet/runtime/pull/57217 as fast as `char` overload? Can `ldstr "."` (`var s = "."`) be optimized as fast as `ldc.i4.s 46` (`char c = '.'`)?

**Comment by svick** [comment]
@ufcpp As far as I can tell, it can. In [my simplified test code](https://sharplab.io/#v2:C4LghgzgtgPgAgJgIwFgBQcAMACOSB0ASgK4B2wAllAKb4DCA9lAA4UA21ATgMpcBuFAMbUIAbnTpE2OugDe6bItwAWaQAsAFAEoFS+WiWHsASXJdmDNmGDUAJt2CcKpAOYAJMKVsdO2NZ+8ubABebFJqAHdtcQMjRX8vH3wAQWZmai8AGQobTjA2DQByfEKtGMMAX11FapVsB05tWv04xVNciysbe0dndwCfPwGg0PCostrDBMDOFLSM22zc/I0AInxVidjFKrRd9E5qADNsCEdiQWATM05O6zsGvo9Erjla7mZPAB5Bf04APmwAH1fmBOGJas4rkCLBDtthagBtACy1GAagYtmMLAKqPRmOxzDYAHlmJQGKQIPgAHIMUxsZx9LQAXVqcFUAHFOAwIgAVNQZRjMACej1cGjwOD4+WI1C02Fku0MKLRGKxOI0eLVhJJZIoFKptPpjNcLLZnO5fIFpCFwrofw0oN80rYsvliqRWoJGq96qJpPJlJpdNIDNITNZ8PZ2FS6S8YpcABEKIdLhKkFKZXKFUqlJ7Vd6iZqC37dYGqckXC5DhAIBQ+NRjeHTZHDHAAMx1WMLJZcFaS7Aut3NSZKCgnDRD2iZDIudEhUJIHTwwwtVqGD7fJ2Ap0QELA3flddj8jYWH7mEMOHHscTjTEKFac9fbD3x+7/Az1zo5c3xRrv8lF3RFYWZfcp0RTBmSPQDFEvPdQnPABqbAkBg49cxvag2AgahRziADYK5Hl+UFBgRQTScsy2QDMOPOAAHZ0MqCQVyUbt416Vxk1TYAqNdOV0Lo/N8VLYtRJ1AN9SDStqxEOsGybCM2U7aMOMWHI+wKJ1B2okc2MUTdSB+P4dz+BCD3M5iTyuc9Qng6zFHHV833IJ8r2wF9XOALQPy/Oc1F/VpCPXYDQPArNHKMeD92Q1CorowxsNw/CFVSwxiKtMiRXtMF+LdBLagqQx0AqIA===), both versions end up with the same code for the fast path and using `mov word ptr [edx+eax*2], 0x2e` for the `.`.

**Comment by jnm2** [comment]
Does that mean the recently-created https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1834 is now obsolete?

![image](https://user-images.githubusercontent.com/8040367/129451389-5f7534d0-bbb5-4fc6-82fc-71427e7a70a6.png)

**Comment by stephentoub** [comment]
No,  this optimization only works due to the call site being a literal. With AppendLiteral, that's expected 100% of the time with normal usage.  That's not the case for StringBuilder.

**Comment by stephentoub** [comment]
Actually. I take that back.  It's already aggressive in lining. It's possible that just works now.
